### PR TITLE
#21272 json system fields split PR

### LINKED ISSUE #21272: Contentlet as Json include system fields
In Order to centralize the JSON generation, We have decided that the new immutable contetlet that is getting stored into the DB from now on must include all system fields. 

The system fields must not be stored as their raw representation they have to have more info so they look as they are hydrated when retrieved by our APIs

- For example, an image field should include the resource link. 
- Binary fields must include additional info from the Metadata at least SHA-256 and Mime-Type. So they can be located regardless of their storage medium.
- Categories must be stored as an array.

### PR DESCRIPTION
**Pull request by fabrizzio-dotCMS**
This is "take two" splitting the PR that adds system fields to the contentlet JSON representation.
This adds a mechanism that allows injecting additional attributes into the contentlet-json "hydrating" them. 
Now we need to re-order the way we save stuff inside internal check-in. Stuff needs to be ready at the moment we save the content into the DB. Files are written until the end. Now we're gonna need them prior to saving if want to be able to generate stuff like metadata. Other stuff we're thinking we're going to include are categories, tags, and the binaries themselves.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/content/business/ContentletJsonAPIImpl.java [hunk 1]
```diff
@@ -140,7 +145,25 @@ public ContentletJsonAPIImpl() {
...
+        Set<Field> categoriesAndTagFields = builder.addAll(contentType.fields(TagField.class))
+                .addAll(contentType.fields(CategoryField.class)).build();
+
+        final Map<String, Object> map = copy.getMap();
+        //A small precaution that guarantees categories and tags will be stored as Lists.
+        categoriesAndTagFields.forEach(field -> {
+            final Object object = map.get(field.variable());
+            if(object instanceof String){
+                final String asString = (String)object;
+                map.put(field.variable(), Arrays.asList(asString.split("\\s*,\\s*")));
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/business/ContentletJsonAPIImpl.java:162]: I would call here something such as 

new StringAsListTransformer().transform(asString)

File: dotCMS/src/integration-test/java/com/dotcms/content/business/ContentletJsonAPITest.java [hunk 2]
```diff
@@ -275,4 +308,185 @@ public void Text_Field_Initialize_Test() throws Exception {
...
+                    .setProperty("title", "lol")
+                    .setProperty("hostFolder", folder.getIdentifier())
+                    .setProperty("imageField",imageFileAsset.getIdentifier())
+                    .nextPersisted();
+
+            assertNotNull(in);
+            final ContentletJsonAPI impl = APILocator.getContentletJsonAPI();
+            final String json = impl.toJson(in);
+            assertNotNull(json);
+            System.out.println(json);
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/content/business/ContentletJsonAPITest.java:341]: I think this System.out can be removed

File: dotCMS/src/main/java/com/dotcms/content/business/ContentletJsonAPIImpl.java [hunk 3]
```diff
@@ -265,12 +310,27 @@ ImmutableContentlet toImmutable(
...
 
         return map;
     }
 
+    /**
+     * These basically tells what system fields are accepted in the generate contentlet-json
+     * @param field
+     * @return
+     */
+    private boolean isAllowedSystemField(final Field field){
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/business/ContentletJsonAPIImpl.java:330]: I am wondering if this should be on the ContentTypeAPI

**Review by nollymar — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/content/business/ContentletJsonAPIImpl.java (hunk 1 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/content/business/ContentletJsonAPIImpl.java:162]: no need to new-up another middleman object. Do we really have something like that?? I would have added that to a utility class and made it static accessible.

File: dotCMS/src/main/java/com/dotcms/content/business/ContentletJsonAPIImpl.java (hunk 3 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/content/business/ContentletJsonAPIImpl.java:330]: I don't think so, it's private logic

File: dotCMS/src/main/java/com/dotcms/content/business/ContentletJsonAPIImpl.java [hunk 4]
```diff
@@ -366,29 +426,112 @@ private boolean isNotMappable(final Field field) {
...
-    private Optional<FieldValue<?>> getFieldValue(final Object value, final Field field) {
-        return field.fieldValue(value);
+    private Optional<FieldValue<?>> hydrateThenGetFieldValue(final Object value, final Field field,
+            final com.dotmarketing.portlets.contentlet.model.Contentlet contentlet) {
+
+        Optional<FieldValueBuilder> fieldValueBuilder = field.fieldValue(value);
+        if (fieldValueBuilder.isPresent()) {
+            FieldValueBuilder builder = fieldValueBuilder.get();
+            final List<Tuple2<HydrationDelegate,String>> delegateAndFields = getHydrationDelegatesFromAnnotations(builder.getClass());
+            for (Tuple2<HydrationDelegate,String> delegateAndField : delegateAndFields) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/business/ContentletJsonAPIImpl.java:458]: set to final

File: dotCMS/src/main/java/com/dotcms/content/model/annotation/HydrateWith.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package com.dotcms.content.model.annotation;
+
+import com.dotcms.content.model.hydration.HydrationDelegate;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+@Retention(RetentionPolicy.RUNTIME)
+public @interface HydrateWith {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/model/annotation/HydrateWith.java:8]: Doc me

File: dotCMS/src/main/java/com/dotcms/content/model/hydration/HydrationDelegate.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package com.dotcms.content.model.hydration;
+
+import com.dotcms.content.model.FieldValueBuilder;
+import com.dotcms.contenttype.model.field.Field;
+import com.dotmarketing.exception.DotDataException;
+import com.dotmarketing.exception.DotSecurityException;
+import com.dotmarketing.portlets.contentlet.model.Contentlet;
+
+public interface HydrationDelegate {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/model/hydration/HydrationDelegate.java:9]: doc me

File: dotCMS/src/main/java/com/dotcms/contenttype/model/field/TagField.java [hunk 7]
```diff
@@ -50,5 +55,23 @@ public DataTypes dataType(){
...
-	
+
+	/**
+	 * {@inheritDoc}
+	 * @return
+	 */
+	@Override
+	public Optional<FieldValueBuilder> fieldValue(Object value) {
+		if (value instanceof String) {
+			final String[] strings = value.toString().split(",");
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/model/field/TagField.java:66]: use StringPool.COMMA

File: dotCMS/src/main/java/com/dotcms/util/ReflectionUtils.java [hunk 8]
```diff
@@ -160,4 +160,22 @@ public static final Object newInstance (final String className) {
...
     }
 
+    /**
+     * Java Beans set value utility
+     * The method d
+     * @param target any object
+     * @param method any method that takes one single param it doesnt have to be named setSomething
+     * @param value The param expected by the method we intent to fire
+     */
+    public static void setValue(final Object target, final String method,final Object value) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/util/ReflectionUtils.java:170]: good, do you have a test?

**Review by jdotcms — APPROVED** [review]
Great Job

File: dotCMS/src/main/java/com/dotcms/util/ReflectionUtils.java (hunk 8 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/util/ReflectionUtils.java:170]: nope but the value being set is covered by the other test.

File: dotCMS/src/main/java/com/dotcms/content/business/ContentletJsonAPIImpl.java (hunk 4 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/content/business/ContentletJsonAPIImpl.java:458]: done

File: dotCMS/src/main/java/com/dotcms/content/model/annotation/HydrateWith.java (hunk 5 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/content/model/annotation/HydrateWith.java:12]: done

File: dotCMS/src/main/java/com/dotcms/content/model/hydration/HydrationDelegate.java (hunk 6 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/content/model/hydration/HydrationDelegate.java:14]: done

File: dotCMS/src/main/java/com/dotcms/content/business/ContentletJsonAPIImpl.java [hunk 9]
```diff
@@ -366,29 +426,112 @@ private boolean isNotMappable(final Field field) {
...
      * @return
      */
-    private Optional<FieldValue<?>> getFieldValue(final Object value, final Field field) {
-        return field.fieldValue(value);
+    private Optional<FieldValue<?>> hydrateThenGetFieldValue(final Object value, final Field field,
+            final com.dotmarketing.portlets.contentlet.model.Contentlet contentlet) {
+
+        Optional<FieldValueBuilder> fieldValueBuilder = field.fieldValue(value);
+        if (fieldValueBuilder.isPresent()) {
+            FieldValueBuilder builder = fieldValueBuilder.get();
```
[the remark this answers is not in the document]
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/content/business/ContentletJsonAPIImpl.java:456]: done

**Pull request merged**